Edit and Remove Buttons from User Settings Page

Jcobo314 opened the pull request:
This patch moves the ability to edit global button definitions from the button panel to the user settings page. This patch also adds functionality to be able to remove a single button definition or clear the entire list. This patch also ensures a button definition cannot be saved with the same name as a previous definition.

Testing Steps:
  1. Create a new session
  2. Click "Settings" in the top left and open the user settings page
  3. Ensure your Button Definition Settings is empty (If not, click "Remove Definition", click "Clear all Button Definitions", and click "Yes". Skip steps 19-20)
  4. Exit the Settings page
  5. Click "+" and create a new button with the name "New Button" and click "Yes" when prompted to save the button definition
  6. Open the Settings page
  7. Ensure the New Button definition has populated in the list
  8. Exit the Settings page
  9. Repeat step 5 with a new hotkey
  10. Open the Settings page
  11. Ensure a second "New Button" definition has NOT populated in the list
  12. Click "Remove Definition"
  13. Enter "New Button" into the field and click "Remove Button Definition"
  14. Ensure the button is deleted from the list immediately
  15. Exit the Settings Page
  16. Create and save two new buttons with different names and hotkeys
  17. Open the settings page and ensure your definitions were saved
  18. Click "Remove Definition"
  19. Click "Clear all Button Definitions" and click "Yes" when prompted
  20. Ensure all button definitions are removed from the list immediately

Type: New Feature

Commit:
Edit Button Commit #1

Commit:
Edit Button Commit #1

Commit:
Button Creation Bug Fix Commit #2

Jcobo314 committed:
Merge pull request #42

Fixes Button Creation Bug

Commit:
Edit Button Commit #2

Commit:
Edit Button Commit #2

Commit:
Merge remote-tracking branch 'origin/dev_button_editing' into dev_button_editing

# Conflicts:
#	Controllers/window_controller.py
#	View/edit_coding_assistance_button_dialog.py

Commit:
Edit Button Commit #3

Commit:
Edit Button Commit #3

Commit:
Merge remote-tracking branch 'origin/dev_button_editing' into dev_button_editing

Briggs-Richardson commented:
1. The button definitions in the settings window should be in a scrollable viewport. I see that the entire settings page is in a scrollable viewport, which is fine. But the button definitions themselves should be in their own scrollable viewport. That way the user doesn't have to scroll down the length of their entire list of button definitions to get to the other settings. By placing the button definitions in their own scrollable viewport, we can control the size of the list in the window.

2. Upon deleting or editing a button definition, it would be convenient to the user if the view of the definition list updated in real time. Right now, the user must close out and reopen the dialog to see the changes. That might be confusing to the user, such that they might not know this is how it works - and may think that their action did not do anything. (Look at the session list in the welcome screen to see what I'm talking about. When a session is removed, it gets deleted from the view immediately.)

3. Since we are using button names to identify buttons to remove or edit, it would make sense if they were forced to be distinct. If the user has two button definitions with the same name, it would not be possible for them to specify which of the two to remove or edit.

Commit:
Edit Button Commit #4

Jcobo314 committed:
Load a Single Button with a Hotkey

This patch changes the UI of the load coding assistance button dialog to ensure a user can only load a single button at a time and allow the user to assign a new hotkey to the button loaded.

Testing Steps:
  1. Create a new session
  2. Load a video file
  3. Click "+" then create and save a new button with hotkey "a" assigned
  4. Close the dialog and ensure the hotkey activates the button created
  5. Click "+" then create and save a new button with no hotkey
  6. Click "Load Button"
  7. Ensure that you can only select one of the two buttons in the list
  8. Select the first button and assign hotkey "b"
  9. Click "Load Button" and close the dialogs until back to the main panel.
  10. Ensure that the button loaded has the same name as the first and that hotkey "a" activates the first button and 
         hotkey "b" activates the second button

Type: Bug Fix

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -609,6 +683,59 @@ def delete_coding_assistance_button(self):
+                self.global_settings_manager.remove_button_definition(button_definition)
+                self.global_settings_manager.save_encoding_button_definitions()
+
+
```
Remove empty line.

Briggs-Richardson commented on View/button_definition_list_element.py:
```diff
@@ -0,0 +1,26 @@
+            column_number = data_index + 1
+            button_definition_layout.addWidget(QLabel("Column " + str(column_number) + " Data: " + data))
+
+        self.setLayout(button_definition_layout)
```
Add a new line at the end of the file.

Briggs-Richardson commented on View/edit_coding_assistance_button_dialog.py:
```diff
@@ -0,0 +1,60 @@
+        """
+        Constructor: Initializes the layout of the Edit Coding Assistance Button Dialog
+        """
+
```
Remove empty line.

Briggs-Richardson commented on View/edit_coding_assistance_button_dialog.py:
```diff
@@ -0,0 +1,60 @@
+        """
+        Connect an edit_button event to a slot function in the controller.
+        """
+        self.edit_button.clicked.connect(slot)
```
Add new line at the end of the file.

Briggs-Richardson commented on View/remove_button_definition_dialog.py:
```diff
@@ -0,0 +1,43 @@
+        """
+        Constructor: Initializes the layout of the Remove Button Definition Dialog
+        """
+
```
Remove empty line.

Briggs-Richardson commented on View/remove_button_definition_dialog.py:
```diff
@@ -0,0 +1,43 @@
+        """
+        Connect a clear all button event to slot
+        """
+        self.clear_all_button.clicked.connect(slot)
```
Add new line at the end of the file.

Briggs-Richardson commented on View/select_edit_button_dialog.py:
```diff
@@ -0,0 +1,36 @@
+        """
+        Constructor: Initializes the layout of the Select Edit Button dialog
+        """
+
```
Remove empty line.

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -621,15 +748,17 @@ def dynamic_button_click(self, button_definition):
         split_one = video_timestamp.split("Time: ")
         split_two = split_one[1].split("/")
         video_timestamp = split_two[0]
+        timestamp_text = QTableWidgetItem(video_timestamp)
```
An exception is thrown every time the user attempts to activate a button when a video isn't loaded. This does not crash the program, but it should be fixed. At the beginning of the method body, we can add a conditional check to make sure that the video has been loaded. If it hasn't been loaded, we can immediately return. Otherwise, we can continue the method execution.

We can most likely check if a video is loaded by checking the source of the media player. If it is None, then a video has not been loaded. Otherwise, it has been loaded.

Briggs-Richardson reviewed (commented):
CR

Commit:
Edit Button Commit #5 (Requested Changes Met)

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -617,19 +743,24 @@ def dynamic_button_click(self, button_definition):
+                        column += 1
+                        data += 1
+                    return
+        else:
```
else: return is redundant, and can be removed.

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -617,19 +743,24 @@ def dynamic_button_click(self, button_definition):
-                    self._window.table_panel.table.setItem(row, column, insert_text)
-                    column += 1
-                return
+        if self._media_player.hasVideo():
```
Instead of wrapping the logic of the entire method in this if-statement, we can change it to the following:

```
if not self._media_player.hasVideo():
    return
# rest of function
```

Briggs-Richardson reviewed (commented):
CR #2

Commit:
Edit Button Commit #6 (Requested Changes Met)
